File: extension/test/hardcoreLock.test.js
// 하드코어 모드가 막아야 하는 변경 목록을 고정한다.
//
// 이 목록이 이 파일의 존재 이유다. 설정 항목이 하나 늘 때마다 "잠가야 하는가"를 묻지 않으면
// 예전처럼 조용히 우회로가 생긴다 - 실제로 화이트리스트·긴급 횟수·예약 차단 삭제·Shorts 항상
// 차단 넷이 그렇게 열려 있었다(documents/QA_REVIEW.md §1.2).

import test from 'node:test';
import assert from 'node:assert/strict';

import {
  HARDCORE_VIOLATION,
  findHardcoreViolations,
  isHardcoreChangeAllowed
} from '../src/lib/hardcoreLock.js';

const BASE = Object.freeze({
  hardcore_mode: true,
  daily_limit_ms: 1800000,
  daily_limit_by_day: { 0: 3600000 },
  shorts_limit_ms: 600000,
  whitelist: ['youtube.com/@lecture'],
  always_block_shorts: true,
  emergency_config: { dailyUses: 3, resetFrequency: 'weekly' },
  scheduled_blocks: [
    { id: 'a', days: [1, 1, 1, 1, 1, 1, 1], startMinute: 1320, endMinute: 420, enabled: true }
  ]
});

const withChange = (patch) => ({ ...BASE, ...patch });
const keysOf = (violations) => violations.map((v) => v.messageKey);

test('하드코어가 꺼져 있으면 무엇이든 허용된다', () => {
  const off = { ...BASE, hardcore_mode: false };
  const result = isHardcoreChangeAllowed(off, {
    ...off,
    daily_limit_ms: 99999999,
    whitelist: ['/']
  });
  assert.equal(result.allowed, true);
  assert.deepEqual(result.violations, []);
});

test('아무것도 안 바꾸면 위반이 없다', () => {
  assert.deepEqual(findHardcoreViolations(BASE, { ...BASE }), []);
});

test('한도를 늘리는 건 막고 줄이는 건 허용한다', () => {
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, withChange({ daily_limit_ms: 3600000 }))), [
    HARDCORE_VIOLATION.dailyLimit
  ]);
  assert.deepEqual(findHardcoreViolations(BASE, withChange({ daily_limit_ms: 600000 })), []);
  // 무제한(0)은 가장 느슨한 값이라 막힌다.
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, withChange({ daily_limit_ms: 0 }))), [
    HARDCORE_VIOLATION.dailyLimit
  ]);
});

test('요일별 한도는 요일 하나만 늘어나도 막는다', () => {
  const loosened = withChange({ daily_limit_by_day: { 0: 7200000 } });
  assert.ok(keysOf(findHardcoreViolations(BASE, loosened)).includes(HARDCORE_VIOLATION.byDayLimit));
  // 요일 오버라이드를 지우면 기본 한도(30분)로 돌아가므로 오히려 세진다 - 허용.
  assert.deepEqual(findHardcoreViolations(BASE, withChange({ daily_limit_by_day: {} })), []);
});

test('Shorts 한도를 늘리거나 푸는 건 막는다', () => {
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, withChange({ shorts_limit_ms: 900000 }))), [
    HARDCORE_VIOLATION.shortsLimit
  ]);
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, withChange({ shorts_limit_ms: 0 }))), [
    HARDCORE_VIOLATION.shortsLimit
  ]);
});

test('화이트리스트는 추가만 막고 삭제는 허용한다', () => {
  const added = withChange({ whitelist: ['youtube.com/@lecture', '@other'] });
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, added)), [HARDCORE_VIOLATION.whitelist]);
  assert.deepEqual(findHardcoreViolations(BASE, withChange({ whitelist: [] })), []);
});

test('긴급 시청 횟수 상향과 리셋 주기 단축을 막는다', () => {
  assert.ok(
    keysOf(
      findHardcoreViolations(
        BASE,
        withChange({ emergency_config: { dailyUses: 5, resetFrequency: 'weekly' } })
      )
    ).includes(HARDCORE_VIOLATION.emergencyUses)
  );
  // weekly -> daily는 "더 자주 채워짐"이라 약화다.
  assert.ok(
    keysOf(
      findHardcoreViolations(
        BASE,
        withChange({ emergency_config: { dailyUses: 3, resetFrequency: 'daily' } })
      )
    ).includes(HARDCORE_VIOLATION.emergencyUses)
  );
  // 줄이는 방향은 허용.
  assert.deepEqual(
    findHardcoreViolations(
      BASE,
      withChange({ emergency_config: { dailyUses: 1, resetFrequency: 'monthly' } })
    ),
    []
  );
});

test('Shorts 항상 차단은 끄는 것만 막는다', () => {
  assert.deepEqual(
    keysOf(findHardcoreViolations(BASE, withChange({ always_block_shorts: false }))),
    [HARDCORE_VIOLATION.alwaysBlockShorts]
  );
  const off = { ...BASE, always_block_shorts: false };
  assert.deepEqual(findHardcoreViolations(off, { ...off, always_block_shorts: true }), []);
});

test('예약 차단은 삭제·비활성화·축소를 막고 추가·확대는 허용한다', () => {
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, withChange({ scheduled_blocks: [] }))), [
    HARDCORE_VIOLATION.scheduledBlocks
  ]);

  const disabled = withChange({
    scheduled_blocks: [{ ...BASE.scheduled_blocks[0], enabled: false }]
  });
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, disabled)), [
    HARDCORE_VIOLATION.scheduledBlocks
  ]);

  const shrunk = withChange({
    scheduled_blocks: [{ ...BASE.scheduled_blocks[0], startMinute: 1380 }]
  });
  assert.deepEqual(keysOf(findHardcoreViolations(BASE, shrunk)), [
    HARDCORE_VIOLATION.scheduledBlocks
  ]);

  const addedBlock = withChange({
    scheduled_blocks: [
      BASE.scheduled_blocks[0],
      { id: 'b', days: [1, 0, 0, 0, 0, 0, 0], startMinute: 600, endMinute: 700, enabled: true }
    ]
  });
  assert.deepEqual(findHardcoreViolations(BASE, addedBlock), []);
});

test('하드코어 해제 요청 자체는 막지 않는다', () => {
  // 해제는 1시간 쿨다운(lib/hardcore.js)이 담당하는 별개 관문이다. 여기서 또 막으면 해제 요청을
  // 저장할 수조차 없게 된다.
  const requested = withChange({ hardcore_disable_requested_at: new Date().toISOString() });
  assert.equal(isHardcoreChangeAllowed(BASE, requested).allowed, true);
});

test('여러 항목을 한 번에 약화시키면 전부 보고한다', () => {
  const all = withChange({
    daily_limit_ms: 3600000,
    whitelist: ['youtube.com/@lecture', '@other'],
    always_block_shorts: false,
    scheduled_blocks: []
  });
  const keys = keysOf(findHardcoreViolations(BASE, all));
  assert.ok(keys.includes(HARDCORE_VIOLATION.dailyLimit));
  assert.ok(keys.includes(HARDCORE_VIOLATION.whitelist));
  assert.ok(keys.includes(HARDCORE_VIOLATION.alwaysBlockShorts));
  assert.ok(keys.includes(HARDCORE_VIOLATION.scheduledBlocks));
});

File: android/app/src/test/java/com/tubelimiter/app/limit/HardcoreLockTest.kt
package com.tubelimiter.app.limit

import com.tubelimiter.app.data.Settings
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

// 규칙의 원본은 확장 `extension/src/lib/hardcoreLock.js`이고, 여기 케이스는
// `extension/test/hardcoreLock.test.js`와 같은 상황을 덮는다 — 한쪽만 바뀌면 같은 계정에서
// 폰과 PC의 판정이 갈라진다(documents/BACKEND.md "하드코어 잠금 범위").

private val ALL_DAYS = List(7) { true }
private val MON_ONLY = listOf(false, true, false, false, false, false, false)

private fun locked(
    dailyLimitMinutes: Int = 30,
    byDayMinutes: List<Int> = List(7) { 30 },
    frequency: LimitFrequency = LimitFrequency.DAILY,
    emergencyAllowance: Int = 3,
    emergencyResetFrequency: EmergencyResetFrequency = EmergencyResetFrequency.DAILY,
    scheduleWindows: List<ScheduleWindow> = emptyList(),
    hardcoreMode: Boolean = true,
) = Settings(
    limit = LimitConfig(dailyLimitMinutes, byDayMinutes, frequency),
    emergencyAllowance = emergencyAllowance,
    emergencyResetFrequency = emergencyResetFrequency,
    scheduleWindows = scheduleWindows,
    hardcoreMode = hardcoreMode,
)

private fun window(
    id: String = "w",
    days: List<Boolean> = ALL_DAYS,
    startMinute: Int = 22 * 60,
    endMinute: Int = 7 * 60,
    enabled: Boolean = true,
) = ScheduleWindow(id, "밤 시간", days, startMinute, endMinute, enabled)

/** 관문이 돌려주는 거부 사유만 꺼내 쓴다 — allowed 플래그는 아래 전용 테스트가 따로 본다. */
private fun violations(previous: Settings, next: Settings): List<HardcoreViolation> =
    isHardcoreChangeAllowed(previous, next).violations

class HardcoreLockTest {

    @Test
    fun `hardcore off means nothing is judged at all`() {
        val previous = locked(dailyLimitMinutes = 15, hardcoreMode = false)
        val next = previous.copy(limit = previous.limit.copy(dailyLimitMinutes = 240))
        assertEquals(emptyList<HardcoreViolation>(), violations(previous, next))
    }

    @Test
    fun `raising the daily limit is refused, lowering it is not`() {
        val previous = locked(dailyLimitMinutes = 30)
        assertEquals(
            listOf(HardcoreViolation.DAILY_LIMIT),
            violations(previous, previous.copy(limit = previous.limit.copy(dailyLimitMinutes = 31))),
        )
        assertEquals(
            emptyList<HardcoreViolation>(),
            violations(previous, previous.copy(limit = previous.limit.copy(dailyLimitMinutes = 29))),
        )
    }

    @Test
    fun `switching the daily limit to unlimited is refused - zero is the loosest value`() {
        val previous = locked(dailyLimitMinutes = 30)
        val toZero = previous.copy(limit = previous.limit.copy(dailyLimitMinutes = 0))
        assertTrue(HardcoreViolation.DAILY_LIMIT in violations(previous, toZero))

        // 반대 방향: 무제한이던 값에 한도를 거는 건 강화라 통과한다.
        val fromZero = locked(dailyLimitMinutes = 0)
        assertEquals(
            emptyList<HardcoreViolation>(),
            violations(fromZero, fromZero.copy(limit = fromZero.limit.copy(dailyLimitMinutes = 30))),
        )
    }

    @Test
    fun `raising a single day of the per-day limit is refused`() {
        val previous = locked(byDayMinutes = listOf(30, 30, 30, 30, 30, 30, 30))
        val next = previous.copy(
            limit = previous.limit.copy(byDayMinutes = listOf(30, 30, 30, 45, 30, 30, 30)),
        )
        assertEquals(listOf(HardcoreViolation.BY_DAY_LIMIT), violations(previous, next))
    }

    @Test
    fun `the unlimited sentinel counts as the loosest per-day value`() {
        val previous = locked(byDayMinutes = List(7) { 30 })
        val next = previous.copy(
            limit = previous.limit.copy(
                byDayMinutes = listOf(30, 30, 30, 30, 30, 30, UNLIMITED_MINUTES),
            ),
        )
        assertTrue(HardcoreViolation.BY_DAY_LIMIT in violations(previous, next))
    }

    @Test
    fun `a per-day list shorter than a week falls back to the daily limit for the missing days`() {
        // 확장이 빠진 요일 키를 기본 한도로 채우는 것과 같은 규칙. 옛 빌드가 남긴 짧은 목록이
        // "판정 대상 없음"으로 조용히 통과하면 안 된다.
        val previous = locked(dailyLimitMinutes = 30, byDayMinutes = listOf(30, 30))
        val next = previous.copy(
            limit = previous.limit.copy(dailyLimitMinutes = 30, byDayMinutes = listOf(30, 30, 90)),
        )
        assertTrue(HardcoreViolation.BY_DAY_LIMIT in violations(previous, next))
    }

    @Test
    fun `switching to the per-day mode is refused when it would raise the real limit`() {
        // 값은 하나도 안 건드리고 모드만 넘기는 우회. 확장에는 없는 안드로이드 전용 항목이다
        // (설정 화면에 모드 칩이 있어서 그 자리가 곧 우회로가 된다).
        val previous = locked(
            dailyLimitMinutes = 15,
            byDayMinutes = List(7) { 120 },
            frequency = LimitFrequency.DAILY,
        )
        val next = previous.copy(limit = previous.limit.copy(frequency = LimitFrequency.BY_DAY))
        assertEquals(listOf(HardcoreViolation.LIMIT_MODE), violations(previous, next))
    }

    @Test
    fun `switching modes is allowed when every day ends up tighter`() {
        val previous = locked(
            dailyLimitMinutes = 60,
            byDayMinutes = List(7) { 60 },
            frequency = LimitFrequency.DAILY,
        )
        val next = previous.copy(
            limit = previous.limit.copy(byDayMinutes = List(7) { 20 }, frequency = LimitFrequency.BY_DAY),
        )
        // 요일별 값이 내려갔으므로 BY_DAY_LIMIT도, 실제 한도가 줄었으므로 LIMIT_MODE도 걸리지 않는다.
        assertEquals(emptyList<HardcoreViolation>(), violations(previous, next))
    }

    @Test
    fun `raising the emergency allowance is refused, lowering it is not`() {
        val previous = locked(emergencyAllowance = 3)
        assertEquals(
            listOf(HardcoreViolation.EMERGENCY),
            violations(previous, previous.copy(emergencyAllowance = 4)),
        )
        assertEquals(
            emptyList<HardcoreViolation>(),
            violations(previous, previous.copy(emergencyAllowance = 2)),
        )
    }

    @Test
    fun `shortening the emergency reset period is refused, lengthening it is not`() {
        // monthly < weekly < daily 순으로 느슨하다 — 자주 리셋될수록 총 허용 횟수가 늘어난다.
        val monthly = locked(emergencyResetFrequency = EmergencyResetFrequency.MONTHLY)
        assertEquals(
            listOf(HardcoreViolation.EMERGENCY),
            violations(monthly, monthly.copy(emergencyResetFrequency = EmergencyResetFrequency.WEEKLY)),
        )
        assertEquals(
            listOf(HardcoreViolation.EMERGENCY),
            violations(monthly, monthly.copy(emergencyResetFrequency = EmergencyResetFrequency.DAILY)),
        )

        val daily = locked(emergencyResetFrequency = EmergencyResetFrequency.DAILY)
        assertEquals(
            emptyList<HardcoreViolation>(),
            violations(daily, daily.copy(emergencyResetFrequency = EmergencyResetFrequency.MONTHLY)),
        )
    }

    @Test
    fun `deleting or disabling a scheduled block is refused`() {
        val previous = locked(scheduleWindows = listOf(window(id = "a")))
        assertEquals(
            listOf(HardcoreViolation.SCHEDULE),
            violations(previous, previous.copy(scheduleWindows = emptyList())),
        )
        assertEquals(
            listOf(HardcoreViolation.SCHEDULE),
            violations(
                previous,
                previous.copy(scheduleWindows = listOf(window(id = "a", enabled = false))),
            ),
        )
    }

    @Test
    fun `shrinking a scheduled block is refused whether by days or by span`() {
        val previous = locked(scheduleWindows = listOf(window(id = "a")))
        val fewerDays = previous.copy(scheduleWindows = listOf(window(id = "a", days = MON_ONLY)))
        assertEquals(listOf(HardcoreViolation.SCHEDULE), violations(previous, fewerDays))

        // 22:00~07:00(9시간)을 23:00~07:00(8시간)으로 줄이는 것.
        val shorterSpan = previous.copy(scheduleWindows = listOf(window(id = "a", startMinute = 23 * 60)))
        assertEquals(listOf(HardcoreViolation.SCHEDULE), violations(previous, shorterSpan))
    }

    @Test
    fun `widening or adding a scheduled block is allowed`() {
        val previous = locked(scheduleWindows = listOf(window(id = "a", days = MON_ONLY)))
        val widened = previous.copy(scheduleWindows = listOf(window(id = "a", days = ALL_DAYS)))
        assertEquals(emptyList<HardcoreViolation>(), violations(previous, widened))

        val added = previous.copy(
            scheduleWindows = listOf(window(id = "a", days = MON_ONLY), window(id = "b")),
        )
        assertEquals(emptyList<HardcoreViolation>(), violations(previous, added))
    }

    @Test
    fun `a wrapping window's blocked span is measured across midnight`() {
        // 22:00~07:00은 wrap을 펴면 540분이다. 09:00~17:00(480분)으로 바꾸면 시각만 보면
        // "늦게 시작해 늦게 끝난다"지만 실제 막는 분량은 줄어들므로 거부돼야 한다.
        val previous = locked(scheduleWindows = listOf(window(id = "a", startMinute = 22 * 60, endMinute = 7 * 60)))
        val next = previous.copy(
            scheduleWindows = listOf(window(id = "a", startMinute = 9 * 60, endMinute = 17 * 60)),
        )
        assertEquals(listOf(HardcoreViolation.SCHEDULE), violations(previous, next))
    }

    @Test
    fun `requesting the hardcore turn-off is never blocked here`() {
        // 해제는 1시간 쿨다운이라는 별개 관문이 담당한다. 여기서 또 막으면 요청 자체를 저장할 수 없다.
        val previous = locked()
        assertEquals(
            emptyList<HardcoreViolation>(),
            violations(previous, previous.copy(hardcoreDisableRequestedAt = 1_700_000_000_000L)),
        )
        assertEquals(
            emptyList<HardcoreViolation>(),
            violations(previous, previous.copy(hardcoreMode = false)),
        )
    }

    @Test
    fun `several weakenings in one save are all reported`() {
        val previous = locked(
            dailyLimitMinutes = 30,
            emergencyAllowance = 1,
            scheduleWindows = listOf(window(id = "a")),
        )
        val next = previous.copy(
            limit = previous.limit.copy(dailyLimitMinutes = 120),
            emergencyAllowance = 5,
            scheduleWindows = emptyList(),
        )
        assertEquals(
            listOf(HardcoreViolation.DAILY_LIMIT, HardcoreViolation.EMERGENCY, HardcoreViolation.SCHEDULE),
            violations(previous, next),
        )
    }

    @Test
    fun `the allowed flag and the violation list always agree`() {
        val previous = locked(dailyLimitMinutes = 30)
        val weakening = isHardcoreChangeAllowed(
            previous,
            previous.copy(limit = previous.limit.copy(dailyLimitMinutes = 120)),
        )
        assertFalse(weakening.allowed)
        assertTrue(weakening.violations.isNotEmpty())

        val tightening = isHardcoreChangeAllowed(
            previous,
            previous.copy(limit = previous.limit.copy(dailyLimitMinutes = 10)),
        )
        assertTrue(tightening.allowed)
        assertTrue(tightening.violations.isEmpty())
    }

    @Test
    fun `every violation names a settings column this client actually owns`() {
        // 거부 사유가 브라우저 전용 컬럼을 가리키면 안 된다 — 안드로이드는 whitelist /
        // always_block_shorts / shorts_limit_ms를 읽지도 쓰지도 않는다.
        val owned = com.tubelimiter.app.sync.SETTINGS_COLUMNS
        HardcoreViolation.entries.forEach { violation ->
            assertTrue(violation.name, violation.column in owned)
        }
    }
}
